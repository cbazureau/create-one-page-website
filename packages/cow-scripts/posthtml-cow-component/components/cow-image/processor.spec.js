/* eslint-disable arrow-body-style */
const { processor, name } = require('./processor');

const workingDir = __dirname;

const mockSharp = jest.fn();

jest.mock('sharp', () => a => {
  return {
    resize: b => {
      return {
        jpeg: c => {
          return { toFile: d => mockSharp(a, b, c, d) };
        },
        png: c => {
          return { toFile: d => mockSharp(a, b, c, d) };
        },
        webp: c => {
          return { toFile: d => !mockSharp(a, b, c, d) };
        },
      };
    },
  };
});

const node = {
  attrs: { classname: 'MyClassName', src: '../mocks/fake.jpg', alt: 'my alt' },
  content: [],
};

beforeEach(() => {
  mockSharp.mockReset();
});

test('cow-image with classname', () => {
  const output = processor(node, { workingDir });
  expect(name).toEqual('cow-image');
  expect(output.attrs.class).toEqual('MyClassName');
  expect(output).toMatchSnapshot();
  expect(
    mockSharp.mock.calls.map(call =>
      call.map((arg, index) =>
        index === 0 || index === 3 ? arg.replace(__dirname, '') : arg
      )
    )
  ).toEqual([
    ['/mocks/fake.jpg', 375, { quality: 80 }, '/mocks/fake.375.jpg'],
    ['/mocks/fake.jpg', 375, { quality: 80 }, '/mocks/fake.375.webp'],
    ['/mocks/fake.jpg', 640, { quality: 80 }, '/mocks/fake.640.jpg'],
    ['/mocks/fake.jpg', 640, { quality: 80 }, '/mocks/fake.640.webp'],
    ['/mocks/fake.jpg', 750, { quality: 80 }, '/mocks/fake.750.jpg'],
    ['/mocks/fake.jpg', 750, { quality: 80 }, '/mocks/fake.750.webp'],
    ['/mocks/fake.jpg', 1024, { quality: 80 }, '/mocks/fake.1024.jpg'],
    ['/mocks/fake.jpg', 1024, { quality: 80 }, '/mocks/fake.1024.webp'],
    ['/mocks/fake.jpg', 1280, { quality: 80 }, '/mocks/fake.1280.jpg'],
    ['/mocks/fake.jpg', 1280, { quality: 80 }, '/mocks/fake.1280.webp'],
    ['/mocks/fake.jpg', 2048, { quality: 80 }, '/mocks/fake.2048.jpg'],
    ['/mocks/fake.jpg', 2048, { quality: 80 }, '/mocks/fake.2048.webp'],
    ['/mocks/fake.jpg', 2560, { quality: 80 }, '/mocks/fake.2560.jpg'],
    ['/mocks/fake.jpg', 2560, { quality: 80 }, '/mocks/fake.2560.webp'],
  ]);
});

test('cow-image without classname', () => {
  const customNode = {
    ...node,
    attrs: { ...node.attrs, classname: undefined },
  };
  const output = processor(customNode, { workingDir });
  expect(output.attrs['data-ref']).toEqual('cow-image');
});

test('cow-image with high priority', () => {
  const customNode = {
    ...node,
    attrs: { ...node.attrs, fetchpriority: 'high' },
  };
  const output = processor(customNode, { workingDir });
  expect(output).toMatchSnapshot();
});

test('cow-image with png', () => {
  const customNode = {
    ...node,
    attrs: { ...node.attrs, src: '../mocks/fake.png' },
  };
  const output = processor(customNode, { workingDir });
  expect(output).toMatchSnapshot();
  expect(
    mockSharp.mock.calls.map(call =>
      call.map((arg, index) =>
        index === 0 || index === 3 ? arg.replace(__dirname, '') : arg
      )
    )
  ).toEqual([
    ['/mocks/fake.png', 375, { compressionLevel: 8 }, '/mocks/fake.375.png'],
    ['/mocks/fake.png', 375, { quality: 80 }, '/mocks/fake.375.webp'],
    ['/mocks/fake.png', 640, { compressionLevel: 8 }, '/mocks/fake.640.png'],
    ['/mocks/fake.png', 640, { quality: 80 }, '/mocks/fake.640.webp'],
    ['/mocks/fake.png', 750, { compressionLevel: 8 }, '/mocks/fake.750.png'],
    ['/mocks/fake.png', 750, { quality: 80 }, '/mocks/fake.750.webp'],
    ['/mocks/fake.png', 1024, { compressionLevel: 8 }, '/mocks/fake.1024.png'],
    ['/mocks/fake.png', 1024, { quality: 80 }, '/mocks/fake.1024.webp'],
    ['/mocks/fake.png', 1280, { compressionLevel: 8 }, '/mocks/fake.1280.png'],
    ['/mocks/fake.png', 1280, { quality: 80 }, '/mocks/fake.1280.webp'],
    ['/mocks/fake.png', 2048, { compressionLevel: 8 }, '/mocks/fake.2048.png'],
    ['/mocks/fake.png', 2048, { quality: 80 }, '/mocks/fake.2048.webp'],
    ['/mocks/fake.png', 2560, { compressionLevel: 8 }, '/mocks/fake.2560.png'],
    ['/mocks/fake.png', 2560, { quality: 80 }, '/mocks/fake.2560.webp'],
  ]);
});
