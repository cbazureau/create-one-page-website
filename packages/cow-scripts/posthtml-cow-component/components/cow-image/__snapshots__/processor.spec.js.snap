// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cow-image with classname 1`] = `
Object {
  "attrs": Object {
    "class": "MyClassName",
    "data-ref": "cow-image",
  },
  "content": Array [
    Object {
      "attrs": Object {
        "data-srcset": "../mocks/fake.375.webp 375w, ../mocks/fake.640.webp 640w, ../mocks/fake.750.webp 750w, ../mocks/fake.1024.webp 1024w, ../mocks/fake.1280.webp 1280w, ../mocks/fake.2048.webp 2048w, ../mocks/fake.2560.webp 2560w",
        "sizes": "100vw",
        "type": "image/webp",
      },
      "tag": "source",
    },
    Object {
      "attrs": Object {
        "data-srcset": "../mocks/fake.375.jpg 375w, ../mocks/fake.640.jpg 640w, ../mocks/fake.750.jpg 750w, ../mocks/fake.1024.jpg 1024w, ../mocks/fake.1280.jpg 1280w, ../mocks/fake.2048.jpg 2048w, ../mocks/fake.2560.jpg 2560w",
        "sizes": "100vw",
        "type": "image/jpeg",
      },
      "tag": "source",
    },
    Object {
      "attrs": Object {
        "alt": "my alt",
        "data-src": "../mocks/fake.375.jpg",
        "src": "../mocks/fake.24.png",
      },
      "tag": "img",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "alt": "my alt",
            "src": "../mocks/fake.375.jpg",
          },
          "tag": "img",
        },
      ],
      "tag": "noscript",
    },
  ],
  "tag": "picture",
}
`;

exports[`cow-image with high priority 1`] = `
Object {
  "attrs": Object {
    "class": "MyClassName",
  },
  "content": Array [
    Object {
      "attrs": Object {
        "sizes": "100vw",
        "srcset": "../mocks/fake.375.webp 375w, ../mocks/fake.640.webp 640w, ../mocks/fake.750.webp 750w, ../mocks/fake.1024.webp 1024w, ../mocks/fake.1280.webp 1280w, ../mocks/fake.2048.webp 2048w, ../mocks/fake.2560.webp 2560w",
        "type": "image/webp",
      },
      "tag": "source",
    },
    Object {
      "attrs": Object {
        "sizes": "100vw",
        "srcset": "../mocks/fake.375.jpg 375w, ../mocks/fake.640.jpg 640w, ../mocks/fake.750.jpg 750w, ../mocks/fake.1024.jpg 1024w, ../mocks/fake.1280.jpg 1280w, ../mocks/fake.2048.jpg 2048w, ../mocks/fake.2560.jpg 2560w",
        "type": "image/jpeg",
      },
      "tag": "source",
    },
    Object {
      "attrs": Object {
        "alt": "my alt",
        "src": "../mocks/fake.375.jpg",
      },
      "fetchpriority": "high",
      "tag": "img",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "alt": "my alt",
            "src": "../mocks/fake.375.jpg",
          },
          "tag": "img",
        },
      ],
      "tag": "noscript",
    },
  ],
  "tag": "picture",
}
`;

exports[`cow-image with png 1`] = `
Object {
  "attrs": Object {
    "class": "MyClassName",
    "data-ref": "cow-image",
  },
  "content": Array [
    Object {
      "attrs": Object {
        "data-srcset": "../mocks/fake.375.webp 375w, ../mocks/fake.640.webp 640w, ../mocks/fake.750.webp 750w, ../mocks/fake.1024.webp 1024w, ../mocks/fake.1280.webp 1280w, ../mocks/fake.2048.webp 2048w, ../mocks/fake.2560.webp 2560w",
        "sizes": "100vw",
        "type": "image/webp",
      },
      "tag": "source",
    },
    Object {
      "attrs": Object {
        "data-srcset": "../mocks/fake.375.png 375w, ../mocks/fake.640.png 640w, ../mocks/fake.750.png 750w, ../mocks/fake.1024.png 1024w, ../mocks/fake.1280.png 1280w, ../mocks/fake.2048.png 2048w, ../mocks/fake.2560.png 2560w",
        "sizes": "100vw",
        "type": "image/png",
      },
      "tag": "source",
    },
    Object {
      "attrs": Object {
        "alt": "my alt",
        "data-src": "../mocks/fake.375.png",
        "src": "../mocks/fake.24.png",
      },
      "tag": "img",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "alt": "my alt",
            "src": "../mocks/fake.375.png",
          },
          "tag": "img",
        },
      ],
      "tag": "noscript",
    },
  ],
  "tag": "picture",
}
`;
